Implementation/Custom-Routing/signalR.md

```diff
@@ -0,0 +1,138 @@
+using Umbraco.Extensions;
+
+public class TestHubRoutes : IAreaRoutes
+{
+    private readonly IRuntimeState _runtimeState;
+    private readonly string _umbracoPathSegment;
+
+    public TestHubRoutes(
+        IOptions<GlobalSettings> globalSettings,
+        IHostingEnvironment hostingEnvironment,
+        IRuntimeState runtimeState)
+    {
+        _runtimeState = runtimeState;
+        _umbracoPathSegment = globalSettings.Value.GetUmbracoMvcArea(hostingEnvironment);
+    }
+
+    public void CreateRoutes(IEndpointRouteBuilder endpoints)
+    {
+        switch (_runtimeState.Level)
+        {
+            case Umbraco.Cms.Core.RuntimeLevel.Run:
+                endpoints.MapHub<TestHub>(GetTestHubRoute());
+                break;
+        }
+
+    }
+
+    public string GetTestHubRoute()
+    {
+        return $"/{_umbracoPathSegment}/{nameof(TestHub)}";
+    }
+}
+```
+
+### Add the routing to the umbraco composer
+Last step in the setup is registering our custom route
+
+```csharp
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
+using Umbraco.Cms.Core.Composing;
+using Umbraco.Cms.Core.DependencyInjection;
+using Umbraco.Cms.Web.Common.ApplicationBuilder;
+using Umbraco.Extensions;
+
+public class TestHubComposer : IComposer
+{
+    public void Compose(IUmbracoBuilder builder)
+    {
+        builder.Services.AddUnique<TestHubRoutes>();
+        builder.Services.AddSignalR();
+        builder.Services.Configure<UmbracoPipelineOptions>(options =>
+        {
+            options.AddFilter(new UmbracoPipelineFilter(
+                "test",
+                applicationBuilder => { },
+                applicationBuilder => { },
+                applicationBuilder =>
+                {
+                    applicationBuilder.UseEndpoints(e =>
+                    {
+                        var hubRoutes = applicationBuilder.ApplicationServices.GetRequiredService<TestHubRoutes>();
+                        hubRoutes.CreateRoutes(e);
+                    });
+                }
+            ));
+        });
+    }
+}
+```
+
+### Testing the setup
+And lastly we can test the setup with some js in our view
+```html
+<!-- We reference the signalR js file that comes with umbraco -->
+<script type="text/javascript" src="/umbraco/lib/signalr/signalr.min.js"></script>
+<script>
+    var connection = new signalR.HubConnectionBuilder()
+        .withUrl("/umbraco/testhub") // this is the url that we created in the routing `TestHubRoutes.GetTestHubRoute()`
+        .withAutomaticReconnect()
+        .configureLogging(signalR.LogLevel.Warning)
+        .build();
+
+    // start the connection
+    connection.start().then(function () {
+        console.info("signalR connection established");
+
+        // connection is established => register our callbacks when the hub sends us something
+        connection.on("Pong", function () {
+            console.log("Pong");
+        });
```

Comment:
You don't have to wait for the connection to be established before you register callbacks, I'd suggest moving this part outside of start, otherwise, you might end up having an absolutely massive start.then() block 😄